File: n-queens.py
from __future__ import print_function

from threading import Thread
from queue import Queue

n = 12

def try_queen(n, board):
    count = 0
    for line in range(n):
        if verify(board + [line]):
            if len(board) + 1 == n:
                count += 1
            else:
                count += try_queen(n, board + [line])
    return count

def verify(board):
    # all on distinct lines
    if len(set(board)) < len(board):
        return False
    last_line, last_col = len(board) - 1, board[-1]
    # all on distinct diagonals
    for line, col in enumerate(board[:-1]):
        if abs(last_line - line) == abs(last_col - col):
            return False
    return True

def run_and_return_to_queue(q, f, *args):
    ret = f(*args)
    q.put(ret)

q = Queue()
threads_list = []

for line in range(n):
    board = [line]
    t = Thread(target=run_and_return_to_queue, args=(q, try_queen, n, board))
    t.start()
    threads_list.append(t)

for t in threads_list:
    t.join()

count = 0
while not q.empty():
    count += q.get()
print(count)

File: n-queens.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

/*
board structure is represented as an array
column = index
line = board[index]
*/

func try_queen(n int, board []int) int {
	count := 0
	for line := 0; line < n; line++ {
		board_new_queen := append(board, line)
		if verify(board_new_queen) {
			if len(board_new_queen) == n {
				count++
			} else {
				count += try_queen(n, board_new_queen)
			}
		}
	}
	return count
}

func verify(board []int) bool {
	last_col := len(board) - 1
	last_line := board[last_col]
	for col, line := range board[:last_col] {
		if line == last_line || abs(last_line-line) == abs(last_col-col) {
			return false
		}
	}
	return true
}

func abs(i int) int {
	if i < 0 {
		return -i
	} else {
		return i
	}
}

func co_try_queen(n int, board []int, count chan int) {
	count <- try_queen(n, board)
}

func main() {
	n := 8
	if len(os.Args) > 1 {
		if i, err := strconv.Atoi(os.Args[1]); err == nil {
			n = i
		}
	}

	count_chan := make(chan int)
	for line := 0; line < n; line++ {
		board := make([]int, 1, n)
		board[0] = line
		go co_try_queen(n, board, count_chan)
	}

	count := 0
	for line := 0; line < n; line++ {
		count += <-count_chan
	}
	fmt.Println(count)
}
